drivers/flash/flash_partition_device.c

```diff
@@ -0,0 +1,205 @@
+	FLASH_API(dev)->GET_ARG_N(1, __VA_ARGS__)(dev, GET_ARGS_LESS_N(1, __VA_ARGS__))
+
+static inline bool is_within_fpd_range(const struct device *dev, off_t offset, size_t len)
+{
+	return (offset >= 0) && ((offset + len) <= FPD_SIZE(dev));
+}
+
+static int fpd_init(const struct device *dev)
+{
+	struct flash_page_info pi;
+	int ret_code = 0;
+	off_t test_offset = FPD_OFFSET(dev, 0);		/* This is offset on real device */
+	const struct flash_parameters *real_dev_params = flash_get_parameters(FPD_REAL_DEV(dev));
+
+	/* Check if device's start offset is aligned with flash page start */
+	ret_code = FLASH_API_CALL(FPD_REAL_DEV(dev), get_page_info, test_offset, &pi);
+	if (ret_code != 0) {
+		/* TODO: Log error, page is not aligned with real device page */
+		return ret_code;
+	}
+
+	FPD_PARAMETERS(dev).erase_value = real_dev_params->erase_value;
+	FPD_PARAMETERS(dev).write_block_size = real_dev_params->write_block_size;
+	FPD_PARAMETERS(dev).flags = 0;
+
+	if (!(real_dev_params->flags & FPF_NON_UNIFORM_LAYOUT)) {
+		const size_t real_max_page_size = real_dev_params->max_page_size;
+
+		if (FPD_SIZE(dev) % real_max_page_size) {
+			/* TODO: Log error, page is not aligned with real device page */
+			return -ERANGE;
+		}
+
+		FPD_PARAMETERS(dev).max_page_size = real_max_page_size;
+		FPD_PAGE_COUNT(dev) = FPD_SIZE(dev) / real_max_page_size;
+
+	} else {
+		/* Check if can read last page of the area */
+		off_t end_offset = test_offset + FPD_SIZE(dev);
+		struct flash_page_info end_pi;
+
+		ret_code = FLASH_API_CALL(FPD_REAL_DEV(dev), get_page_info, end_offset, &end_pi);
+
+		if (ret_code != 0) {
+			/* TODO: Log error, probably passing past the device */
+			return ret_code;
+		}
+		/* Check area size is aligned to last page */
+		if ((test_offset + FPD_SIZE(dev)) != (end_pi.offset  + end_pi.size)) {
+			return -ERANGE;
+		}
+
+		/*
+		 * The first page already read, so there is at least one, prepare offset for next
+		 * one, and it is largest page found so far
+		 */
+		FPD_PAGE_COUNT(dev) = 1;
+		test_offset += pi.size;
+		FPD_PARAMETERS(dev).max_page_size = pi.size;
+
+		while (FPD_OFFSET(dev, test_offset) < FPD_SIZE(dev)) {
+			ret_code = FLASH_API_CALL(FPD_REAL_DEV(dev), get_page_info, test_offset,
+						  &pi);
+
+			if (ret_code != 0) {
+				/* TODO: Log error, page is not aligned with real device page */
+				return ret_code;
+			}
+
+			FPD_PAGE_COUNT(dev)++;
+
+			test_offset = pi.offset + pi.size;
+
+			/*
+			 * If current largest page differs from current page, we have non-uniform
+			 * layout.
+			 */
+			if (FPD_PARAMETERS(dev).max_page_size != pi.size) {
+				FPD_PARAMETERS(dev).flags |= FPF_NON_UNIFORM_LAYOUT;
+			}
+
+			/* New largest page? */
+			if (pi.size > FPD_PARAMETERS(dev).max_page_size) {
+				FPD_PARAMETERS(dev).max_page_size = pi.size;
+			}
+		};
+	}
+
+	return 0;
+}
+
+static int fpd_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	if (is_within_fpd_range(dev, offset, len)) {
```

Comment:
I think the normal convention is to exit early on error (for this function and those below is):
```
if (!is_within_fpd_range(...)) {
    return -EINVAL;
}
return FLASH_API_CALL(...);
```

Reply:
Can do that..